## Setup

### Importació de les dades.

In [2]:
library(ElemStatLearn)
data(spam)
wine<-read.csv('wine.csv',header=FALSE)
colnames(wine)<-c("Type","Alcohol","Malic","Ash","Alcalinity","Magnesium","Phenols","Flavonoids","Nonflavonoids","Proanthocyanins","Color","Hue","Dilution","Proline")
wine$Type <- as.factor(wine$Type)

### Llibreries

In [3]:
library(class)

# Classificació de Spam

Aquest dataset ens classifica entre si es spam o email un correu, segons uns quants parametres.

In [4]:
nrow(spam)

[1] 4601

Veiem que el dataset és bastant gran, aixi que utilitzarem com a model de predicció **k-fold cross-validation**. 
No utilitcem Leave-one-out per que seria poc eficient fer les $4601$ fileres, tampoc utilitzarem boostrap ja que no tenim necessitat de ampliar la mostra.

Utilitzarem un k-fold de 10

In [18]:
#Desordenem el dataset
spam<-spam[sample(nrow(spam)),]

### Classificació per  K-nn
Utilitzarem la funció *knn* que trobem al paquet *class*.
Amb la quantitat de dades que hi han testejarem amb una **k= 10** i **20**.

#### K= 10

In [6]:
k<-10

kfold<-10
totalprecisio.email<-0
totalprecisio.spam<-0
for (i in 1:kfold){
    ITest<-((nrow(spam)/kfold)*(i-1)+1):((nrow(spam)/kfold)*i)
    spam.train<-spam[-ITest,-ncol(spam)]
    spam.test<-spam[ITest,-ncol(spam)]
    spam.Ytrain<-spam[-ITest,ncol(spam)]
    spam.Ytest<-spam[ITest,ncol(spam)]
    spam.Yhat<-knn(spam.train, spam.test, spam.Ytrain, k = k)
    C<-table("True"=spam.Ytest,"Predicted"=spam.Yhat)
    cat('\nMatriu confusio\n')
    print(C)
    precisio.email<-C[1]/(C[1]+C[3])
    precisio.spam<-C[4]/(C[2]+C[4])
    cat('preciso email ',+precisio.email)
    cat('\nprecisio spam ',precisio.spam)
    cat('\n')
    totalprecisio.email<-totalprecisio.email + precisio.email
    totalprecisio.spam<-totalprecisio.spam + precisio.spam
}
cat('\npreciso email mitjana ',totalprecisio.email/kfold)
cat('\nprecisio spam mitjana ',totalprecisio.spam/kfold)


Matriu confusio
       Predicted
True    email spam
  email   231   58
  spam     45  126
preciso email  0.799308
precisio spam  0.7368421

Matriu confusio
       Predicted
True    email spam
  email   221   48
  spam     58  133
preciso email  0.8215613
precisio spam  0.6963351

Matriu confusio
       Predicted
True    email spam
  email   213   49
  spam     59  139
preciso email  0.8129771
precisio spam  0.7020202

Matriu confusio
       Predicted
True    email spam
  email   236   44
  spam     42  138
preciso email  0.8428571
precisio spam  0.7666667

Matriu confusio
       Predicted
True    email spam
  email   248   49
  spam     43  120
preciso email  0.8350168
precisio spam  0.7361963

Matriu confusio
       Predicted
True    email spam
  email   233   38
  spam     55  134
preciso email  0.8597786
precisio spam  0.7089947

Matriu confusio
       Predicted
True    email spam
  email   249   24
  spam     47  140
preciso email  0.9120879
precisio spam  0.7486631

Matriu confus

#### K= 20

In [69]:
k<-20

kfold<-10
totalprecisio.email<-0
totalprecisio.spam<-0
for (i in 1:kfold){
    ITest<-((nrow(spam)/kfold)*(i-1)+1):((nrow(spam)/kfold)*i)
    spam.train<-spam[-ITest,-ncol(spam)]
    spam.test<-spam[ITest,-ncol(spam)]
    spam.Ytrain<-spam[-ITest,ncol(spam)]
    spam.Ytest<-spam[ITest,ncol(spam)]
    spam.Yhat<-knn(spam.train, spam.test, spam.Ytrain, k = k)
    C<-table("True"=spam.Ytest,"Predicted"=spam.Yhat)
    cat('\nMatriu confusio\n')
    print(C)
    precisio.email<-C[1]/(C[1]+C[3])
    precisio.spam<-C[4]/(C[2]+C[4])
    cat('preciso email ',+precisio.email)
    cat('\nprecisio spam ',precisio.spam)
    cat('\n')
    totalprecisio.email<-totalprecisio.email + precisio.email
    totalprecisio.spam<-totalprecisio.spam + precisio.spam
}
cat('\npreciso email mitjana ',totalprecisio.email/kfold)
cat('\nprecisio spam mitjana ',totalprecisio.spam/kfold)


Matriu confusio
       Predicted
True    email spam
  email   221   39
  spam     61  139
preciso email  0.85
precisio spam  0.695

Matriu confusio
       Predicted
True    email spam
  email   244   49
  spam     53  114
preciso email  0.8327645
precisio spam  0.6826347

Matriu confusio
       Predicted
True    email spam
  email   221   69
  spam     51  119
preciso email  0.762069
precisio spam  0.7

Matriu confusio
       Predicted
True    email spam
  email   242   39
  spam     64  115
preciso email  0.86121
precisio spam  0.6424581

Matriu confusio
       Predicted
True    email spam
  email   214   56
  spam     42  148
preciso email  0.7925926
precisio spam  0.7789474

Matriu confusio
       Predicted
True    email spam
  email   242   52
  spam     63  103
preciso email  0.8231293
precisio spam  0.6204819

Matriu confusio
       Predicted
True    email spam
  email   227   47
  spam     49  137
preciso email  0.8284672
precisio spam  0.7365591

Matriu confusio
       Predict

Veiem que optenim mes precisió en $k = 10$ que en $k = 20$. Encara que la precisió de predicció de spam és més baixa que amb email.

### Regresió logística

Com que amb la regressió logística hi podem veure l'error de test no utilitzare **k-fold cross validation**.
Si no que farem un **Hold-out**, on el 10% de les dades sera de test i la restra de train.

Categoricem email i spam a 0 i 1.

In [24]:
x <- c(spam = 0, email = 1)
spam.categorize<-spam[-nrow(spam)]
spam.categorize$spam <- as.numeric(x[spam$spam])
str(spam)

'data.frame':	4601 obs. of  58 variables:
 $ A.1 : num  0 0.49 0.09 0 0 0 0 0 0 0 ...
 $ A.2 : num  0 0 0.19 0 0 0 0 0 0 0 ...
 $ A.3 : num  0 0.99 0.98 0 0 0 0.29 0 0 0 ...
 $ A.4 : num  0.15 0 0 0 0 0 0 0 0 0 ...
 $ A.5 : num  0.46 0 0.78 0 0 0 0.29 0 0 0 ...
 $ A.6 : num  0 0.99 0.78 0 0 0.25 0 0 0 0 ...
 $ A.7 : num  0 0 0.49 0 0 0 0 0 0 0 ...
 $ A.8 : num  0.15 0 0 0 0 1 0 0 0 0 ...
 $ A.9 : num  0 0 0.78 0 0 0 0 0 0 0 ...
 $ A.10: num  0 0.99 0.19 0 0 0 0 0 0 0 ...
 $ A.11: num  0 0.99 0.29 0 0 0 0 0 0 0 ...
 $ A.12: num  0.92 2.48 0.68 0 0 0.25 2.05 0 0 0 ...
 $ A.13: num  0 0.49 0.29 0 0 0 0 0 0 0 ...
 $ A.14: num  0.15 0 0 0 0 0 0 0 0 0 ...
 $ A.15: num  0 0 1.57 0 0 0 0 0 0 0 ...
 $ A.16: num  0 4.97 0.09 0 0 0 0 0 0 0 ...
 $ A.17: num  0.46 0.99 0 0 0 0 0 0 0 0 ...
 $ A.18: num  0 0 1.08 0 0 0.25 0 0 0 0.89 ...
 $ A.19: num  0 3.48 1.28 0 0 0 0 0 0 0.89 ...
 $ A.20: num  0 0 0.19 0 0 0 0 0 0 0 ...
 $ A.21: num  0 1.99 0.68 0 0 0 0 0 0 0 ...
 $ A.22: num  0 0 0 0 0 0 0 0 0 0 

In [25]:
ITest<-sample(1:nrow(spam),nrow(spam)/10,replace=FALSE)
spam.train<-spam.categorize[-ITest,]
spam.test<-spam.categorize[ITest,]
spam.logist<-glm(spam~.,data=spam.train,family=binomial)

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”

In [26]:
summary(spam.logist)
spam.pred<-predict(spam.logist,newdata=spam.test, type="response")
spam.pred.crisp<-1*(spam.pred>=0.5)
C<-table("True"=spam.test$spam,"Predicted"=spam.pred.crisp)
C


Call:
glm(formula = spam ~ ., family = binomial, data = spam.train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-4.8505  -0.1971   0.0000   0.1053   5.1389  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.625e+00  1.528e-01 -10.634  < 2e-16 ***
A.1         -4.324e-01  2.469e-01  -1.751 0.079955 .  
A.2         -1.662e-01  8.344e-02  -1.992 0.046365 *  
A.3          8.943e-02  1.166e-01   0.767 0.442918    
A.4          1.485e+00  1.477e+00   1.005 0.314962    
A.5          5.805e-01  1.065e-01   5.449 5.07e-08 ***
A.6          8.106e-01  2.683e-01   3.021 0.002520 ** 
A.7          2.233e+00  3.352e-01   6.663 2.68e-11 ***
A.8          7.143e-01  2.062e-01   3.465 0.000530 ***
A.9          7.399e-01  2.912e-01   2.541 0.011055 *  
A.10         1.179e-01  7.419e-02   1.589 0.112130    
A.11        -2.401e-01  3.063e-01  -0.784 0.433167    
A.12        -1.214e-01  7.659e-02  -1.585 0.113071    
A.13        -1.159e-01  2.422e-0

    Predicted
True   0   1
   0 274  11
   1  21 154

Els resultats del classificadors son gaire bons però anem veure si treient alguns descriptors podem millorar el clasificador.
Utilitzarem la funció *step* per a fer-ho.

In [ ]:
spam.logist2<-step(spam.logist)

Start:  AIC=1745.11
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.31 + A.32 + A.33 + A.34 + A.35 + A.36 + 
    A.37 + A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + 
    A.46 + A.47 + A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + 
    A.55 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: algorithm did not converge”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted 

       Df Deviance   AIC
- A.34  1     1629  1743
- A.31  1     1629  1743
- A.37  1     1629  1743
- A.55  1     1629  1743
- A.13  1     1629  1743
- A.18  1     1629  1743
- A.51  1     1630  1744
- A.32  1     1630  1744
- A.3   1     1630  1744
- A.11  1     1630  1744
- A.50  1     1630  1744
- A.30  1     1630  1744
- A.14  1     1631  1745
<none>        1629  1745
- A.22  1     1631  1745
- A.10  1     1632  1746
- A.12  1     1632  1746
- A.38  1     1632  1746
- A.15  1     1632  1746
- A.1   1     1632  1746
- A.47  1     1633  1747
- A.19  1     1634  1748
- A.54  1     1634  1748
- A.43  1     1634  1748
- A.39  1     1635  1749
- A.9   1     1636  1750
- A.26  1     1636  1750
- A.28  1     1637  1751
- A.36  1     1637  1751
- A.4   1     1637  1751
- A.2   1     1637  1751
- A.29  1     1638  1752
- A.33  1     1638  1752
- A.24  1     1640  1754
- A.35  1     1640  1754
- A.6   1     1641  1755
- A.56  1     1641  1755
- A.48  1     1642  1756
- A.20  1     1644  1758


Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1743.18
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.31 + A.32 + A.33 + A.35 + A.36 + A.37 + 
    A.38 + A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + 
    A.47 + A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.55 + 
    A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: algorithm did not converge”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted 

       Df Deviance   AIC
- A.31  1     1629  1741
- A.37  1     1629  1741
- A.55  1     1629  1741
- A.13  1     1629  1741
- A.18  1     1630  1742
- A.51  1     1630  1742
- A.32  1     1630  1742
- A.3   1     1630  1742
- A.11  1     1630  1742
- A.50  1     1630  1742
- A.30  1     1630  1742
- A.14  1     1631  1743
<none>        1629  1743
- A.22  1     1631  1743
- A.10  1     1632  1744
- A.12  1     1632  1744
- A.38  1     1632  1744
- A.15  1     1632  1744
- A.1   1     1633  1745
- A.47  1     1633  1745
- A.19  1     1634  1746
- A.54  1     1634  1746
- A.43  1     1634  1746
- A.39  1     1635  1747
- A.9   1     1636  1748
- A.26  1     1636  1748
- A.28  1     1637  1749
- A.36  1     1637  1749
- A.4   1     1637  1749
- A.2   1     1637  1749
- A.29  1     1638  1750
- A.33  1     1638  1750
- A.24  1     1640  1752
- A.35  1     1640  1752
- A.6   1     1641  1753
- A.56  1     1641  1753
- A.48  1     1642  1754
- A.20  1     1644  1756
- A.49  1     1646  1758


Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1741.33
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.32 + A.33 + A.35 + A.36 + A.37 + A.38 + 
    A.39 + A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + 
    A.48 + A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.55 + A.56 + 
    A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.37  1   1629.5 1739.5
- A.55  1   1629.5 1739.5
- A.13  1   1629.6 1739.6
- A.18  1   1629.7 1739.7
- A.51  1   1629.8 1739.8
- A.32  1   1629.9 1739.9
- A.3   1   1629.9 1739.9
- A.40  1   1630.0 1740.0
- A.11  1   1630.0 1740.0
- A.50  1   1630.0 1740.0
- A.30  1   1630.3 1740.3
- A.14  1   1630.7 1740.7
<none>      1629.3 1741.3
- A.22  1   1631.6 1741.6
- A.10  1   1632.0 1742.0
- A.12  1   1632.0 1742.0
- A.38  1   1632.4 1742.4
- A.15  1   1632.5 1742.5
- A.1   1   1632.7 1742.7
- A.47  1   1633.6 1743.6
- A.19  1   1633.8 1743.8
- A.54  1   1634.4 1744.4
- A.43  1   1634.4 1744.4
- A.39  1   1635.6 1745.6
- A.9   1   1636.3 1746.3
- A.26  1   1636.5 1746.5
- A.28  1   1637.2 1747.2
- A.36  1   1637.3 1747.3
- A.4   1   1637.4 1747.4
- A.2   1   1637.6 1747.6
- A.29  1   1638.5 1748.5
- A.33  1   1638.5 1748.5
- A.24  1   1640.1 1750.1
- A.35  1   1640.3 1750.3
- A.6   1   1640.8 1750.8
- A.56  1   1641.0 1751.0
- A.48  1   1642.4 1752.4
- A.20  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1739.5
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + 
    A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + 
    A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.55 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.55  1   1629.7 1737.7
- A.13  1   1629.8 1737.8
- A.18  1   1629.8 1737.8
- A.51  1   1630.0 1738.0
- A.32  1   1630.0 1738.0
- A.3   1   1630.1 1738.1
- A.11  1   1630.1 1738.1
- A.40  1   1630.1 1738.1
- A.50  1   1630.2 1738.2
- A.30  1   1630.5 1738.5
- A.14  1   1630.9 1738.9
<none>      1629.5 1739.5
- A.22  1   1631.7 1739.7
- A.10  1   1632.1 1740.1
- A.12  1   1632.2 1740.2
- A.38  1   1632.6 1740.6
- A.15  1   1632.7 1740.7
- A.1   1   1632.8 1740.8
- A.47  1   1633.8 1741.8
- A.19  1   1633.9 1741.9
- A.43  1   1634.5 1742.5
- A.54  1   1634.6 1742.6
- A.39  1   1635.8 1743.8
- A.9   1   1636.4 1744.4
- A.26  1   1636.6 1744.6
- A.28  1   1637.2 1745.2
- A.36  1   1637.4 1745.4
- A.4   1   1637.5 1745.5
- A.2   1   1637.9 1745.9
- A.33  1   1638.6 1746.6
- A.29  1   1638.7 1746.7
- A.24  1   1640.2 1748.2
- A.35  1   1640.4 1748.4
- A.6   1   1641.0 1749.0
- A.56  1   1641.1 1749.1
- A.48  1   1642.5 1750.5
- A.20  1   1644.2 1752.2
- A.49  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1737.73
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.13 + A.14 + A.15 + A.16 + A.17 + A.18 + 
    A.19 + A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + 
    A.28 + A.29 + A.30 + A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + 
    A.40 + A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + 
    A.49 + A.50 + A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.13  1   1630.0 1736.0
- A.18  1   1630.0 1736.0
- A.51  1   1630.2 1736.2
- A.32  1   1630.3 1736.3
- A.3   1   1630.3 1736.3
- A.40  1   1630.4 1736.4
- A.50  1   1630.4 1736.4
- A.11  1   1630.4 1736.4
- A.30  1   1630.7 1736.7
- A.14  1   1631.1 1737.1
<none>      1629.7 1737.7
- A.22  1   1632.0 1738.0
- A.10  1   1632.5 1738.5
- A.12  1   1632.5 1738.5
- A.38  1   1632.8 1738.8
- A.15  1   1632.9 1738.9
- A.1   1   1633.0 1739.0
- A.47  1   1634.0 1740.0
- A.19  1   1634.2 1740.2
- A.43  1   1634.7 1740.7
- A.54  1   1634.8 1740.8
- A.39  1   1636.0 1742.0
- A.9   1   1636.4 1742.4
- A.26  1   1636.9 1742.9
- A.28  1   1637.5 1743.5
- A.36  1   1637.6 1743.6
- A.4   1   1637.7 1743.7
- A.2   1   1638.2 1744.2
- A.29  1   1638.9 1744.9
- A.33  1   1639.0 1745.0
- A.24  1   1640.5 1746.5
- A.6   1   1641.1 1747.1
- A.35  1   1641.4 1747.4
- A.48  1   1642.6 1748.6
- A.20  1   1644.3 1750.3
- A.49  1   1646.0 1752.0
- A.57  1   1649.0 1755.0
- A.41  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1736
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.18 + A.19 + 
    A.20 + A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + 
    A.29 + A.30 + A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + 
    A.41 + A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + 
    A.50 + A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.18  1   1630.3 1734.3
- A.51  1   1630.5 1734.5
- A.3   1   1630.5 1734.5
- A.32  1   1630.6 1734.6
- A.50  1   1630.6 1734.6
- A.40  1   1630.6 1734.6
- A.11  1   1630.6 1734.6
- A.30  1   1630.9 1734.9
- A.14  1   1631.3 1735.3
<none>      1630.0 1736.0
- A.22  1   1632.3 1736.3
- A.12  1   1632.8 1736.8
- A.10  1   1632.8 1736.8
- A.15  1   1633.0 1737.0
- A.38  1   1633.1 1737.1
- A.1   1   1633.2 1737.2
- A.47  1   1634.2 1738.2
- A.19  1   1634.4 1738.4
- A.43  1   1635.0 1739.0
- A.54  1   1635.2 1739.2
- A.39  1   1636.4 1740.4
- A.9   1   1636.7 1740.7
- A.26  1   1637.1 1741.1
- A.28  1   1637.8 1741.8
- A.36  1   1638.0 1742.0
- A.4   1   1638.0 1742.0
- A.2   1   1638.4 1742.4
- A.33  1   1639.1 1743.1
- A.29  1   1639.2 1743.2
- A.24  1   1640.7 1744.7
- A.6   1   1641.2 1745.2
- A.35  1   1641.6 1745.6
- A.48  1   1642.9 1746.9
- A.20  1   1644.5 1748.5
- A.49  1   1646.3 1750.3
- A.57  1   1649.3 1753.3
- A.41  1   1650.2 1754.2
- A.8   1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1734.27
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.30 + A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + 
    A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.50 + 
    A.51 + A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.51  1   1630.8 1732.8
- A.3   1   1630.8 1732.8
- A.32  1   1630.9 1732.9
- A.11  1   1630.9 1732.9
- A.50  1   1630.9 1732.9
- A.40  1   1630.9 1732.9
- A.30  1   1631.2 1733.2
- A.14  1   1631.6 1733.6
<none>      1630.3 1734.3
- A.22  1   1632.6 1734.6
- A.10  1   1633.0 1735.0
- A.12  1   1633.1 1735.1
- A.38  1   1633.4 1735.4
- A.1   1   1633.7 1735.7
- A.15  1   1633.7 1735.7
- A.47  1   1634.4 1736.4
- A.19  1   1634.9 1736.9
- A.43  1   1635.3 1737.3
- A.54  1   1635.4 1737.4
- A.39  1   1636.8 1738.8
- A.9   1   1637.0 1739.0
- A.26  1   1637.3 1739.3
- A.28  1   1638.1 1740.1
- A.36  1   1638.2 1740.2
- A.4   1   1638.3 1740.3
- A.2   1   1638.6 1740.6
- A.33  1   1639.5 1741.5
- A.29  1   1639.6 1741.6
- A.24  1   1640.9 1742.9
- A.6   1   1641.3 1743.3
- A.35  1   1642.0 1744.0
- A.48  1   1643.2 1745.2
- A.20  1   1644.7 1746.7
- A.49  1   1646.7 1748.7
- A.57  1   1649.4 1751.4
- A.41  1   1650.3 1752.3
- A.8   1   1651.1 1753.1
- A.44  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1732.75
spam ~ A.1 + A.2 + A.3 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + 
    A.10 + A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + 
    A.21 + A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + 
    A.30 + A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + 
    A.42 + A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.50 + 
    A.52 + A.53 + A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.3   1   1631.3 1731.3
- A.32  1   1631.3 1731.3
- A.11  1   1631.3 1731.3
- A.50  1   1631.4 1731.4
- A.40  1   1631.4 1731.4
- A.30  1   1631.7 1731.7
- A.14  1   1632.1 1732.1
<none>      1630.8 1732.8
- A.22  1   1633.1 1733.1
- A.12  1   1633.5 1733.5
- A.10  1   1633.5 1733.5
- A.38  1   1633.9 1733.9
- A.1   1   1634.1 1734.1
- A.15  1   1634.2 1734.2
- A.47  1   1634.8 1734.8
- A.19  1   1635.5 1735.5
- A.43  1   1635.8 1735.8
- A.54  1   1635.8 1735.8
- A.39  1   1637.2 1737.2
- A.9   1   1637.5 1737.5
- A.26  1   1637.8 1737.8
- A.28  1   1638.6 1738.6
- A.36  1   1638.8 1738.8
- A.4   1   1638.8 1738.8
- A.2   1   1638.9 1738.9
- A.33  1   1640.0 1740.0
- A.29  1   1640.1 1740.1
- A.24  1   1641.4 1741.4
- A.6   1   1641.8 1741.8
- A.35  1   1642.4 1742.4
- A.48  1   1643.6 1743.6
- A.20  1   1645.2 1745.2
- A.49  1   1647.1 1747.1
- A.57  1   1649.8 1749.8
- A.41  1   1650.7 1750.7
- A.8   1   1651.7 1751.7
- A.44  1   1651.7 1751.7
- A.21  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1731.34
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + 
    A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.30 + 
    A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + 
    A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.50 + A.52 + 
    A.53 + A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.50  1   1631.9 1729.9
- A.40  1   1632.0 1730.0
- A.32  1   1632.0 1730.0
- A.11  1   1632.0 1730.0
- A.30  1   1632.2 1730.2
- A.14  1   1632.6 1730.6
<none>      1631.3 1731.3
- A.22  1   1633.7 1731.7
- A.12  1   1633.9 1731.9
- A.10  1   1634.0 1732.0
- A.38  1   1634.5 1732.5
- A.1   1   1634.8 1732.8
- A.15  1   1634.9 1732.9
- A.47  1   1635.2 1733.2
- A.19  1   1636.3 1734.3
- A.43  1   1636.4 1734.4
- A.54  1   1636.4 1734.4
- A.39  1   1637.7 1735.7
- A.9   1   1638.2 1736.2
- A.26  1   1638.4 1736.4
- A.28  1   1639.1 1737.1
- A.36  1   1639.3 1737.3
- A.4   1   1639.3 1737.3
- A.2   1   1639.7 1737.7
- A.29  1   1640.4 1738.4
- A.33  1   1640.9 1738.9
- A.24  1   1642.0 1740.0
- A.6   1   1642.3 1740.3
- A.35  1   1643.1 1741.1
- A.48  1   1644.2 1742.2
- A.20  1   1645.8 1743.8
- A.49  1   1647.8 1745.8
- A.57  1   1650.4 1748.4
- A.41  1   1651.3 1749.3
- A.8   1   1652.0 1750.0
- A.44  1   1652.7 1750.7
- A.21  1   1654.9 1752.9
- A.17  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1729.94
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + 
    A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.30 + 
    A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.40 + A.41 + A.42 + 
    A.43 + A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.52 + A.53 + 
    A.54 + A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.40  1   1632.5 1728.5
- A.32  1   1632.5 1728.5
- A.11  1   1632.6 1728.6
- A.30  1   1632.8 1728.8
- A.14  1   1633.2 1729.2
<none>      1631.9 1729.9
- A.22  1   1634.4 1730.4
- A.12  1   1634.5 1730.5
- A.10  1   1634.6 1730.6
- A.38  1   1635.1 1731.1
- A.1   1   1635.4 1731.4
- A.15  1   1635.6 1731.6
- A.47  1   1635.9 1731.9
- A.43  1   1637.0 1733.0
- A.54  1   1637.0 1733.0
- A.19  1   1637.3 1733.3
- A.39  1   1638.2 1734.2
- A.9   1   1638.8 1734.8
- A.26  1   1639.0 1735.0
- A.28  1   1639.1 1735.1
- A.36  1   1639.9 1735.9
- A.4   1   1640.0 1736.0
- A.2   1   1640.0 1736.0
- A.29  1   1641.0 1737.0
- A.33  1   1641.4 1737.4
- A.24  1   1642.5 1738.5
- A.6   1   1642.8 1738.8
- A.35  1   1643.6 1739.6
- A.48  1   1644.8 1740.8
- A.20  1   1646.5 1742.5
- A.49  1   1648.5 1744.5
- A.57  1   1650.9 1746.9
- A.41  1   1651.8 1747.8
- A.8   1   1652.8 1748.8
- A.44  1   1653.2 1749.2
- A.21  1   1655.9 1751.9
- A.17  1   1659.2 1755.2
- A.56  1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1728.54
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + 
    A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.30 + 
    A.32 + A.33 + A.35 + A.36 + A.38 + A.39 + A.41 + A.42 + A.43 + 
    A.44 + A.45 + A.46 + A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + 
    A.56 + A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

       Df Deviance    AIC
- A.32  1   1633.1 1727.1
- A.11  1   1633.2 1727.2
- A.30  1   1633.4 1727.4
- A.14  1   1633.8 1727.8
<none>      1632.5 1728.5
- A.12  1   1635.0 1729.0
- A.10  1   1635.1 1729.1
- A.22  1   1635.3 1729.3
- A.38  1   1635.7 1729.7
- A.1   1   1635.9 1729.9
- A.15  1   1636.1 1730.1
- A.47  1   1636.5 1730.5
- A.54  1   1637.2 1731.2
- A.43  1   1637.6 1731.6
- A.19  1   1638.1 1732.1
- A.39  1   1638.8 1732.8
- A.9   1   1639.2 1733.2
- A.26  1   1639.5 1733.5
- A.28  1   1639.7 1733.7
- A.36  1   1640.5 1734.5
- A.2   1   1640.5 1734.5
- A.4   1   1640.5 1734.5
- A.29  1   1641.5 1735.5
- A.33  1   1641.9 1735.9
- A.24  1   1643.1 1737.1
- A.6   1   1643.4 1737.4
- A.35  1   1644.2 1738.2
- A.48  1   1645.5 1739.5
- A.20  1   1647.1 1741.1
- A.49  1   1649.1 1743.1
- A.57  1   1651.6 1745.6
- A.41  1   1652.4 1746.4
- A.8   1   1653.5 1747.5
- A.44  1   1653.6 1747.6
- A.21  1   1656.3 1750.3
- A.17  1   1659.6 1753.6
- A.56  1   1663.6 1757.6
- A.5   1   

Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”


Step:  AIC=1727.14
spam ~ A.1 + A.2 + A.4 + A.5 + A.6 + A.7 + A.8 + A.9 + A.10 + 
    A.11 + A.12 + A.14 + A.15 + A.16 + A.17 + A.19 + A.20 + A.21 + 
    A.22 + A.23 + A.24 + A.25 + A.26 + A.27 + A.28 + A.29 + A.30 + 
    A.33 + A.35 + A.36 + A.38 + A.39 + A.41 + A.42 + A.43 + A.44 + 
    A.45 + A.46 + A.47 + A.48 + A.49 + A.52 + A.53 + A.54 + A.56 + 
    A.57



Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning message:
“glm.fit: fitted probabilities numerically 0 or 1 occurred”Warning mess

In [ ]:
summary(spam.logist2)
spam.pred<-predict(spam.logist2,newdata=spam.test, type="response")
spam.pred.crisp<-1*(spam.pred>=0.5)
C<-table("True"=spam.test$spam,"Predicted"=spam.pred.crisp)
C

# Wine

In [18]:
nrow(wine)

[1] 178

Veiem que hi ha molt pocs elementsal dataset. Haurem d'utilitzar metodes de validació com el **Leave-one-out** o **Bootstrap** que permeten tenir un conjunt de train el més gran possible.


## Discriminador lineal de Fisher

In [ ]:
library(MASS)

In [9]:
for (i in 1:10){
    n<- nrow(wine)
    iTrain<-sample(1:n,n,replace = TRUE)
    wine.train<-wine[iTrain,]
    wine.test<-wine[-iTrain,]
    wine.lda1<-lda(Type~.,data=wine.train)
    wine.pred<-predict(wine.lda1,newdata=wine.test,type="response")
    C<-table("True"=wine.test$Type,"Predicted"=wine.pred$class)
    cat("\n length iTest ", nrow(wine.test),'\n')
    print(C)
    cat("\n")
}


 length iTest  68 
    Predicted
True  1  2  3
   1 24  0  0
   2  1 26  0
   3  0  1 16


 length iTest  63 
    Predicted
True  1  2  3
   1 19  1  0
   2  0 26  1
   3  0  0 16


 length iTest  68 
    Predicted
True  1  2  3
   1 27  0  0
   2  1 21  0
   3  0  1 18


 length iTest  65 
    Predicted
True  1  2  3
   1 20  0  0
   2  1 25  1
   3  0  0 18


 length iTest  67 
    Predicted
True  1  2  3
   1 23  0  0
   2  0 26  0
   3  0  0 18


 length iTest  67 
    Predicted
True  1  2  3
   1 24  0  0
   2  0 25  1
   3  0  1 16


 length iTest  66 
    Predicted
True  1  2  3
   1 21  0  0
   2  0 30  0
   3  0  0 15


 length iTest  65 
    Predicted
True  1  2  3
   1 19  1  0
   2  0 24  0
   3  0  1 20


 length iTest  63 
    Predicted
True  1  2  3
   1 19  0  0
   2  2 28  1
   3  0  0 13


 length iTest  73 
    Predicted
True  1  2  3
   1 28  0  0
   2  0 25  2
   3  0  0 18



# Conctrete

In [12]:
require(regsel)
data(concrete)

conc.s<-scale(concrete)
str(conc.s)

conc.s.df<-as.data.frame(conc.s)
str(conc.s.df)

 num [1:1030, 1:9] 2.477 2.477 0.491 0.491 -0.79 ...
 - attr(*, "dimnames")=List of 2
  ..$ : NULL
  ..$ : chr [1:9] "Cement" "BlastFurnaceSlag" "FlyAsh" "Water" ...
 - attr(*, "scaled:center")= Named num [1:9] 281.2 73.9 54.2 181.6 6.2 ...
  ..- attr(*, "names")= chr [1:9] "Cement" "BlastFurnaceSlag" "FlyAsh" "Water" ...
 - attr(*, "scaled:scale")= Named num [1:9] 104.51 86.28 64 21.35 5.97 ...
  ..- attr(*, "names")= chr [1:9] "Cement" "BlastFurnaceSlag" "FlyAsh" "Water" ...
'data.frame':	1030 obs. of  9 variables:
 $ Cement             : num  2.477 2.477 0.491 0.491 -0.79 ...
 $ BlastFurnaceSlag   : num  -0.856 -0.856 0.795 0.795 0.678 ...
 $ FlyAsh             : num  -0.847 -0.847 -0.847 -0.847 -0.847 ...
 $ Water              : num  -0.916 -0.916 2.174 2.174 0.489 ...
 $ Superplasticizer   : num  -0.62 -0.62 -1.04 -1.04 -1.04 ...
 $ CoarseAggregate    : num  0.8627 1.0557 -0.5263 -0.5263 0.0705 ...
 $ FineAggregate      : num  -1.217 -1.217 -2.24 -2.24 0.648 ...
 $ Age            

In [14]:
conc.ls1<-lm(Cement~BlastFurnaceSlag+FlyAsh+Water+Superplasticizer+CoarseAggregate+FineAggregate+Age+CompressiveStrength, data=conc.s.df)
summary(conc.ls1)


Call:
lm(formula = Cement ~ BlastFurnaceSlag + FlyAsh + Water + Superplasticizer + 
    CoarseAggregate + FineAggregate + Age + CompressiveStrength, 
    data = conc.s.df)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.11042 -0.22021 -0.03327  0.18994  1.76766 

Coefficients:
                      Estimate Std. Error t value Pr(>|t|)    
(Intercept)         -6.559e-16  1.046e-02   0.000    1.000    
BlastFurnaceSlag    -8.505e-01  1.303e-02 -65.278   <2e-16 ***
FlyAsh              -7.413e-01  1.301e-02 -56.967   <2e-16 ***
Water               -5.279e-01  2.245e-02 -23.515   <2e-16 ***
Superplasticizer    -2.778e-02  1.807e-02  -1.537    0.125    
CoarseAggregate     -5.540e-01  1.602e-02 -34.578   <2e-16 ***
FineAggregate       -7.155e-01  1.637e-02 -43.724   <2e-16 ***
Age                 -1.139e-01  1.276e-02  -8.929   <2e-16 ***
CompressiveStrength  2.178e-01  1.543e-02  14.113   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual s

In [15]:
step(conc.ls1)

Start:  AIC=-2240.39
Cement ~ BlastFurnaceSlag + FlyAsh + Water + Superplasticizer + 
    CoarseAggregate + FineAggregate + Age + CompressiveStrength

                      Df Sum of Sq    RSS      AIC
<none>                             114.97 -2240.39
- Superplasticizer     1      0.27 115.24 -2240.01
- Age                  1      8.98 123.95 -2164.95
- CompressiveStrength  1     22.43 137.40 -2058.83
- Water                1     62.27 177.24 -1796.61
- CoarseAggregate      1    134.64 249.62 -1443.91
- FineAggregate        1    215.28 330.25 -1155.58
- FlyAsh               1    365.44 480.42  -769.54
- BlastFurnaceSlag     1    479.85 594.82  -549.52



Call:
lm(formula = Cement ~ BlastFurnaceSlag + FlyAsh + Water + Superplasticizer + 
    CoarseAggregate + FineAggregate + Age + CompressiveStrength, 
    data = conc.s.df)

Coefficients:
        (Intercept)     BlastFurnaceSlag               FlyAsh  
         -6.559e-16           -8.505e-01           -7.413e-01  
              Water     Superplasticizer      CoarseAggregate  
         -5.279e-01           -2.778e-02           -5.540e-01  
      FineAggregate                  Age  CompressiveStrength  
         -7.155e-01           -1.139e-01            2.178e-01  
